feat: Add AbstractUuid.title

kulla opened the pull request:
Closes https://github.com/serlo/api.serlo.org/issues/627

hugotiburtino reviewed (changes requested):
As in the issue, in case of threads it should return the title of the parent or empty.
Anyway, if you manually test v.g. with id 15469 (a thread of comment 15468), it throws ` "Error: Cannot return null for non-nullable field Comment.title."`

Besides that, we have to clear with @elbotho if it is enough for exercises to have a title like"ExerciseGroup 2217" and not the name of parent, as asked in issue.

elbotho commented:
Thank you both. 

> ExerciseGroup 2217
here I'd actually prefere the title of the parent or undefined because then I can use at least a translated type string (e.g. "Aufgabengruppe").

Having a title field in abstact.uuid is a great improvement anyway, but for the entities without own titles (theads, comments, Exercises…) we'll probably can't avoid to do some logic in the frontend.

kulla commented:
> I'd actually prefere the title of the parent or undefined because then I can use at least a translated type string (e.g. "Aufgabengruppe").

Implemented

> As in the issue, in case of threads it should return the title of the parent or empty.

Implemented as well. @elbotho When there is an answer of the thread the title of the first comment in the thread shall returned, right?

elbotho commented:
> the thread the title of the first comment in the thread shall returned,

yes, but that should also be the title of the parent of the thread, right?

LarsTheGlidingSquirrel commented:
I had an idea to make this more transparent, maybe: 
API sends AbstractUuid.title and AbstractUuid.parentTitle. 

Some entities have an obvious title (Pages, Courses, ...). Some do not (Solution, GroupedExercise, ...), but for these we might want to have the title of the context they belong to (Folder, Course, ...). In the frontend (at least where the notifications are displayed, the rest I don't know) we could then make a simple check: If title is set -> Display title. Else -> Display localized string for entity type ("Exercise", "Solution", ...) and then parentTitle ("Exercises about topic XY", ...)

elbotho commented:
>API sends AbstractUuid.title and AbstractUuid.parentTitle.

thanks @LarsTheGlidingSquirrel! Thought about this as well since the current solution then needs implicit knowlege (but is also simple and short).

> Display localized string for entity type ("Exercise", "Solution", ...) and then parentTitle ("Exercises about topic XY", ...)

My guess is that the frontend at some point will have to check the type anyway and generate the title accordingly because of small differences like:
- "Exercises for ${(parent)title}"
- "Thread on ${(parent)title}"

----
How I see it now:

Pros of only `title`:
- title could be `String!` and we don't need an extra check if it exists
- only one extra field

Pros of `title` and `parentTitle`
- defined clearer, needs less implicit knowledge

My main goal for this is an easy way to display a link to an entity without a crazy long query.
So I'll let @kulla and @hugotiburtino decide how they want to implement it :)
I don't have a strong opinion.

kulla commented:
Okay, I ended up with the following solution: The frontend can display the link to an uuid with `${i18n(uuid.___typename)} ${uuid.title}`. When a title of an uuid is not present the title of the parent is returned. See https://github.com/serlo/api.serlo.org/blob/627-add-title-field-to-uuid/packages/server/src/schema/uuid/abstract-uuid/utils.ts#L71-L109 @hugotiburtino Ready to be merged
